### DISCUSSION TITLE
what is the entrance point of anaconda?

### DISCUSSION THREAD

**Discussion by jshen28** [rhinstaller/anaconda#3549] [Q&A] [answered]
Hello community!

I am newbie and trying to figure out how does centos kickstart work.  I am able to boot a virtual machine using `kickstart` config file. Now I am trying to figure out how does kickstart config got interpreted and executed. 

Is anaconda a auto start service which got executed when iso boots? If so, what is the program that gets executed?

Sorry if this has already be answered...

Best,
Norman

**Comment by VladimirSlavik** [comment, marked as answer]
Hi Norman! In most cases, we have an environment that we call `boot.iso`. That is a self-contained system which has one purpose, to start anaconda. We have some `dracut` hooks to do stuff as the system is starting, so technically there is some "anaconda" executed even "before the system". Later, `systemd` starts `anaconda.service`, which has some dependencies ([see our `systemd` files](https://github.com/rhinstaller/anaconda/tree/master/data/systemd)) and runs `anaconda.py` (the toplevel one) within a `tmux` session. Anaconda then decides if GUI is needed and starts `Xorg` (or not). From there, you are "in the installer" as you know it as an user.

As to kickstart... During system start in `dracut` ([see dir dracut in repo](https://github.com/rhinstaller/anaconda/tree/master/dracut)), `%pre` scripts run, so that might be interpreting too? Then, once anaconda code actually starts, multiple modules (processes) are started and register which kickstart commands they process. Kickstart is "assembled" (includes etc.) and then redistributed to the modules, and from there it's no longer kickstart as text but rather data structures.

There are other modes of running anaconda, but they mostly forbid or restrict use of kickstart, so they are probably not as relevant to your question.

(Feel free to ask for more details.)

**Reply by VladimirSlavik** [reply]
PS: If you want to browse anaconda's python code, using an IDE such as Pycharm could be a good idea.

Starting at anaconda.py works to some extent. That file contains a lot of the starting-up things, but there is a point where execution leaves the linear model and goes into a loop (abstracted away in some class), and then you'll probably get lost. (I did...)

For the rest, it's probably better to check the GUI files in `pyanaconda/ui/gui/spokes` and see what module stuff they manipulate. Modules are the separate processes and have code under `pyanaconda/modules/*/`. There's an additional layer (`dasbus` - another repo) that translates Python calls to D-Bus and back between UI and modules, since these are different processes. Fortunately `dasbus` is mostly transparent to the IDE and the code appears as if in one process when you read it.

**Reply by jshen28** [reply]
Thank you very much for the reply! How does a installer iso get created? For example, am I able to create myself an iso from a running server and it behaves like a normal centos iso?

**Reply by VladimirSlavik** [reply]
Yes, that should be possible. You have to run [`lorax`](https://github.com/weldr/lorax) to make `boot.iso`. I don't know it enough to give some definitive advice, but:
- Lorax has [docs](https://weldr.io/lorax/), so you can read them, same as I would. Unfortunately they are not well structured, or just a tough read,... or I'm dumb :-)
- Lorax can mess up a system very thoroughly, best use a disposable VM or containers with enough privileges to try the builds.
- Lorax is very finicky about loop devices, be prepared to reboot or manually add them multiple times. Each run can consume two. If you're doing stuff in a container, tough - loop devices are controlled by the host. Does not apply to a VM
- It is easiest to start by building an image for a released (GA) fedora version. Most docs will mention rawhide, but that can be broken on its own, which is bad for learning. I don't recall seeing anything about CentOS. But the principles should be the same.
- We build boot.iso for kickstart tests every time we run the tests, if you know Github actions and dockerfiles, you can piece it together from these. The actual build "invocation" is in [this script](https://github.com/rhinstaller/anaconda/blob/master/dockerfile/anaconda-iso-creator/lorax-build), but be mindful of all the dependencies resolved by previous steps. Action [kickstart-tests.yml](https://github.com/rhinstaller/anaconda/blob/master/.github/workflows/kickstart-tests.yml) + Dockerfile [anaconda-iso-creator](https://github.com/rhinstaller/anaconda/tree/master/dockerfile/anaconda-iso-creator) + others, start from the action to unravel that pile.

**Reply by jshen28** [reply]
Thank you :)

**Reply by poncovka** [reply]
Btw. if you just want to test your changes, you can use an updates image.

https://fedoraproject.org/wiki/Anaconda/Updates
https://anaconda-installer.readthedocs.io/en/latest/boot-options.html#inst-updates
https://github.com/rhinstaller/devel-tools/tree/master/reanaconda
